packages/backend/src/tenants/service.ts

```diff
@@ -0,0 +1,202 @@
+import { ApolloClient } from '@apollo/client'
+import { TransactionOrKnex } from 'objection'
+
+export interface TenantService {
+  get: (id: string) => Promise<TenantWithIdpConfig | undefined>
+  create: (options: CreateTenantOptions) => Promise<Tenant>
+  update: (options: UpdateTenantOptions) => Promise<Tenant>
+  delete: (id: string) => Promise<void>
+}
+
+export interface ServiceDependencies extends BaseService {
+  knex: TransactionOrKnex
+  apolloClient: ApolloClient<NormalizedCacheObject>
+}
+
+export async function createTenantService(
+  deps_: ServiceDependencies
+): Promise<TenantService> {
+  const deps: ServiceDependencies = {
+    ...deps_,
+    logger: deps_.logger.child({ service: 'TenantService' })
+  }
+
+  return {
+    get: (id: string) => getTenant(deps, id),
+    create: (options) => createTenant(deps, options),
+    update: (options) => updateTenant(deps, options),
+    delete: (id) => deleteTenant(deps, id)
+  }
+}
+
+async function getTenant(
+  deps: ServiceDependencies,
+  id: string
+): Promise<TenantWithIdpConfig | undefined> {
+  const tenant = await Tenant.query(deps.knex).findById(id)
+  if (!tenant) return undefined
+
+  const query = gql`
+    query GetAuthTenant($input: GetTenantInput!) {
+      getTenant(input: $input) {
+        tenant {
+          id
+          idpConsentUrl
+          idpSecret
+        }
+      }
+    }
+  `
+  const variables = { input: { id } }
+  // TODO: add type to this in https://github.com/interledger/rafiki/issues/3125
+  const authTenantResponse = await deps.apolloClient.query({ query, variables })
+  const authTenant = authTenantResponse.data.getTenant.tenant
+  if (!authTenant) {
+    deps.logger.error(
+      { tenantId: id },
+      'could not find auth tenant entry for existing backend entry'
+    )
+    return undefined
+  }
+
+  const { idpConsentUrl, idpSecret } = authTenant
+  return {
+    ...tenant,
+    idpConsentUrl,
+    idpSecret
+  }
+}
+
+interface CreateTenantOptions {
+  email: string
+  apiSecret: string
+  idpSecret: string
+  idpConsentUrl: string
+  publicName?: string
+}
+
+async function createTenant(
+  deps: ServiceDependencies,
+  options: CreateTenantOptions
+): Promise<Tenant> {
+  const trx = await deps.knex.transaction()
+  try {
+    const { email, apiSecret, publicName, idpSecret, idpConsentUrl } = options
+    const tenant = await Tenant.query(trx).insertAndFetch({
+      email,
+      publicName,
+      apiSecret
+    })
+
+    const mutation = gql`
+      mutation CreateAuthTenant($input: CreateTenantInput!) {
+        createTenant(input: $input) {
+          tenant {
+            id
+          }
+        }
+      }
+    `
+
+    const variables = {
+      input: {
+        id: tenant.id,
+        idpSecret,
+        idpConsentUrl
+      }
+    }
+
+    // TODO: add type to this in https://github.com/interledger/rafiki/issues/3125
+    await deps.apolloClient.mutate({ mutation, variables })
```

Comment:
Does mutate method throw error in case of 4xx / 5xx errors ?

Reply:
It does - a manifestation of this is [having to catch `400` responses inside a try/catch block in the tests](https://github.com/interledger/rafiki/pull/3140/files#diff-38f750d470d3a544f415e5642aed59c1fa4f3d899445deea7c84958eaa1af779R107-R123).

Reply:
My understanding is this will be a new gql server for service-to-service communication so I guess maybe we can simply return a 400 from that server like this expects, but I just wanted to note that our current resolvers dont behave this way.

For errors we do something like `throw new GraphQLError('tenant not found', { extensions: { code: 'NOT_FOUND' } })`. I tested an existing one of these on Bruno and see the gql server replies with a 200 and error details in the body like:

```ts
{
  "errors": [
    {
      "message": "No grant",
      "locations": [
        {
          "line": 2,
          "column": 5
        }
      ],
      "path": [
        "grant"
      ],
      "extensions": {
        "code": "NOT_FOUND"
      }
    }
  ],
  "data": null
}
```

Im not necessarily against doing it a different way, if that seems better. I just wanted to note this. 

Also I think the test would incorrectly pass if we updated the mock to something like the above response because the expects in the catch would never be called. We safeguard against this sort of incorrect pass elsewhere by doing `expect.assertions(3)` (or whatever number).

Reply:
Thanks for pointing that out. I updated the nocks to return proper error structure. Mercifully, the client still parses that as an error and throws, but it's a better test to have the correct response from the API.

Reply:
> Mercifully, the client still parses that as an error and throws

Good to know, was curious about that